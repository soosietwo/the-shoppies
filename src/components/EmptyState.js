import React from "react";
import { EmptyState } from "@shopify/polaris";

const EmptyStateMessage = (movies) => {
  return !movies.length ? (
    <EmptyState
      heading="No movies found"
      action=""
      image="https://cdn.shopify.com/s/files/1/2376/3301/products/emptystate-files.png"
    >
      <p>Search to get started!</p>
    </EmptyState>
  ) : undefined;
};

export default EmptyStateMessage;
