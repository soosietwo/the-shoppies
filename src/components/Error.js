import React from "react";
import { Banner } from "@shopify/polaris";

// TODO: make this more user friendly and less scary
const Error = (props) => (
  <Banner status="critical" title={props.title}>
    <p>{props.details}</p>
  </Banner>
);

export default Error;
