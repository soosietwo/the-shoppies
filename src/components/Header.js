import React from "react";
import { TopBar } from "@shopify/polaris";
import SearchBar from "./SearchBar";

import { gql, useQuery } from "@apollo/client";

export const NOMINEES_CONNECTION_QUERY = gql`
  query NOMINEES_CONNECTION_QUERY {
    nomineesConnection {
      aggregate {
        count
      }
    }
  }
`;

const Header = ({ toggleSheetActive, nomineesCount }) => {
  const { loading, error, data } = useQuery(NOMINEES_CONNECTION_QUERY);
  const count = data ? data.nomineesConnection.aggregate.count : 0;

  return (
    <TopBar
      showNavigationToggle
      searchField={<SearchBar />}
      secondaryMenu={
        <TopBar.Menu
          activatorContent={
            <span>
              Nominees
              <span
                style={{
                  padding: "5px 9px",
                  borderRadius: "50%",
                  background: "red",
                  color: "white",
                  marginLeft: "5px",
                  fontSize: "1.2rem",
                }}
              >
                {count}
              </span>
            </span>
          }
          onOpen={toggleSheetActive}
          onClose={toggleSheetActive}
        />
      }
    />
  );
};

export default Header;
