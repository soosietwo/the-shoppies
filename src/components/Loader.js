import React from "react";
import {
  ResourceItem,
  TextContainer,
  SkeletonDisplayText,
  SkeletonBodyText,
  ResourceList,
  SkeletonThumbnail,
  Stack,
} from "@shopify/polaris";

const Loader = ({ count }) => {
  return (
    <ResourceList
      items={[...Array(count).keys()]}
      renderItem={() => (
        <ResourceItem>
          <Stack>
            <SkeletonThumbnail size="large" />
            <Stack.Item fill>
              <TextContainer>
                <SkeletonDisplayText size="small" />
                <SkeletonBodyText />
              </TextContainer>
            </Stack.Item>
          </Stack>
        </ResourceItem>
      )}
    />
  );
};

export default Loader;
