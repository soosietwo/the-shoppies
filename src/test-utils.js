// test-utils.js
import React from "react";
import { render } from "@testing-library/react";
import { AppProvider, Frame } from "@shopify/polaris";
import en from "@shopify/polaris/locales/en.json";

Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});
Object.defineProperty(window, "scroll", { value: jest.fn(), writable: true });

const AllTheProviders = ({ children }) => {
  return (
    <div id="test-bed">
      <AppProvider i18n={en}>
        <Frame>{children}</Frame>
      </AppProvider>
    </div>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
