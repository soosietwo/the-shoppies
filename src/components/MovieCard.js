import React from "react";
import {
  Heading,
  Subheading,
  Button,
  ResourceItem,
  TextContainer,
} from "@shopify/polaris";
import { gql, useMutation } from "@apollo/client";

import { NOMINEES_QUERY } from "./Sidebar";

const ADD_NOMINEE_MUTATION = gql`
  mutation ADD_NOMINEE_MUTATION(
    $id: ID!
    $title: String!
    $poster: String!
    $year: Int!
  ) {
    addNominee(id: $id, title: $title, poster: $poster, year: $year) {
      id
      title
      poster
      year
    }
  }
`;

const MovieCard = (props) => {
  const {
    disableAction,
    actionText,
    action,
    movie: { title, year, poster, id },
  } = props;

  const posterMarkup = (
    <img
      style={{ maxWidth: "100px" }}
      src={poster}
      alt={`Poster for ${title}`}
    />
  );

  const [
    addNominee,
    { loading, data, error },
  ] = useMutation(ADD_NOMINEE_MUTATION, {
    refetchQueries: [{ query: NOMINEES_QUERY }],
  });

  return (
    <ResourceItem
      id={id}
      media={posterMarkup}
      accessibilityLabel={`View details for ${title}`}
    >
      <TextContainer>
        <Heading>{title}</Heading>
        <Subheading>{year}</Subheading>

        {action && (
          <Button
            disabled={disableAction && disableAction(props.movie)}
            onClick={() => {
              const variables = {
                id,
                poster,
                year,
                title,
              };

              addNominee({ variables });
            }}
          >
            {actionText}
          </Button>
        )}
      </TextContainer>
    </ResourceItem>
  );
};

export default MovieCard;
