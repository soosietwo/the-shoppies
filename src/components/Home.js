import React, { useContext, useCallback, useState } from "react";
import { Page, Card, Frame, Banner } from "@shopify/polaris";
import { gql, useQuery } from "@apollo/client";

import MoviesList from "./MoviesList";
import Pagination from "./Pagination";
import Loader from "./Loader";
import Error from "./Error";
import Sidebar from "./Sidebar";
import Header from "./Header";
import MovieCard from "./MovieCard";
import { Context } from "../store";

export const NOMINEES_QUERY = gql`
  query NOMINEES_QUERY {
    nominees {
      id
      title
      poster
      year
    }
  }
`;

const Home = () => {
  const [state] = useContext(Context);
  const [sheetActive, setSheetActive] = useState(false);
  const toggleSheetActive = useCallback(
    () => setSheetActive((sheetActive) => !sheetActive),
    []
  );

  const { loading, error, data } = useQuery(NOMINEES_QUERY);

  return (
    <Frame topBar={<Header toggleSheetActive={toggleSheetActive} />}>
      <Page>
        <Card>
          {state.error && !state.isLoading && (
            <Error title="Oops! Something went wrong" details={state.error} />
          )}

          {data && data.nominees.length === 5 && (
            <Banner
              status="success"
              title="All done!"
              action={{
                content: "Go to nominees",
                onAction: () => toggleSheetActive(true),
              }}
            >
              <p>You have chosen 5 nominees. Submit or remove them.</p>
            </Banner>
          )}

          {state.isLoading || loading ? (
            <Loader count={10} />
          ) : (
            <MoviesList
              singular="movie"
              plural="movies"
              movies={state.movies}
              totalResults={state.totalResults}
              emptyStateTitle="No movies found"
              emptyStateCaption="Start typing in your favourite movie titles to get started!"
              renderItem={(movie) => (
                <MovieCard
                  movie={movie}
                  nominees={(data && data.nominees) || []}
                  toggleSheetActive={toggleSheetActive}
                />
              )}
            />
          )}

          {state.totalResults > 10 && <Pagination />}
        </Card>

        <Sidebar
          nominees={(data && data.nominees) || []}
          toggleSheetActive={toggleSheetActive}
          sheetActive={sheetActive}
        />
      </Page>
    </Frame>
  );
};

export default Home;
