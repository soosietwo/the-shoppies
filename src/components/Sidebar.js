import React, { useState } from "react";
import {
  Sheet,
  Heading,
  Button,
  Scrollable,
  Toast,
  Banner,
} from "@shopify/polaris";
import { MobileCancelMajor } from "@shopify/polaris-icons";
import { gql, useMutation } from "@apollo/client";

import MoviesList from "./MoviesList";
import NomineeCard from "./NomineeCard";
import { NOMINEES_QUERY } from "./Home";
import { NOMINEES_CONNECTION_QUERY } from "./Header";

const DELETE_ALL_NOMINEES_MUTATION = gql`
  mutation DELETE_ALL_NOMINEES_MUTATION {
    deleteAllNominees {
      count
    }
  }
`;

const Sidebar = ({ toggleSheetActive, sheetActive, nominees }) => {
  const [showToast, setShowToast] = useState(false);

  const [deleteAllNominees, { loading }] = useMutation(
    DELETE_ALL_NOMINEES_MUTATION,
    {
      refetchQueries: [
        { query: NOMINEES_QUERY },
        { query: NOMINEES_CONNECTION_QUERY },
      ],
      update(cache, { data, error }) {
        if (error) return console.error(error);
        if (data) setShowToast(true);
      },
      optimisticResponse: {
        __typename: "Mutation",
        deleteAllNominees: {
          count: 5,
          __typename: "BatchPayload",
        },
      },
    }
  );

  return (
    <Sheet open={sheetActive}>
      <div
        style={{
          display: "flex",
          flexDirection: "column",
          height: "100%",
        }}
      >
        <div
          style={{
            alignItems: "center",
            borderBottom: "1px solid #DFE3E8",
            display: "flex",
            justifyContent: "space-between",
            padding: "1.6rem",
            width: "100%",
          }}
        >
          <Heading>Your nominations</Heading>
          <Button
            accessibilityLabel="Cancel"
            icon={MobileCancelMajor}
            onClick={toggleSheetActive}
            plain
          />
        </div>
        {nominees.length === 5 && (
          <Banner status="success" title="All done!">
            <p>You have chosen 5 nominees. Submit or remove them.</p>
          </Banner>
        )}
        <Scrollable style={{ height: "100%" }}>
          <MoviesList
            singular="nominee"
            plural="nominees"
            movies={nominees}
            emptyStateTitle="No nominees yet"
            emptyStateCaption="Choose five of your favourite movies to nominate!"
            renderItem={(movie) => <NomineeCard movie={movie} />}
          />
        </Scrollable>
        <div
          style={{
            alignItems: "center",
            borderTop: "1px solid #DFE3E8",
            display: "flex",
            justifyContent: "space-between",
            padding: "1.6rem",
            width: "100%",
          }}
        >
          <Button loading={loading} onClick={toggleSheetActive}>
            Cancel
          </Button>
          <Button
            disabled={nominees.length !== 5}
            loading={loading}
            primary
            onClick={deleteAllNominees}
          >
            Submit Nominees
          </Button>
        </div>
      </div>
      {showToast ? (
        <Toast
          content="Nominees Submitted!"
          onDismiss={() => setShowToast(false)}
        />
      ) : null}
    </Sheet>
  );
};

export default Sidebar;
