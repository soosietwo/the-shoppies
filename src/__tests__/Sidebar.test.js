import React from "react";
import { render, waitFor } from "../test-utils";
import { MockedProvider } from "@apollo/client/testing";
import Sidebar from "../components/Sidebar";
import { NOMINEES_QUERY } from "../components/Home";

const mockNominee = {
  id: "132",
  title: "test",
  poster: "123",
  year: 1234,
};

const mocks = [
  {
    request: { query: NOMINEES_QUERY },
    result: {
      data: {
        nominees: [mockNominee],
      },
    },
  },
];

describe("Sidebar", () => {
  it("renders and matches snapshot", async () => {
    const { baseElement } = render(
      <MockedProvider mocks={mocks}>
        <Sidebar sheetActive={true} nominees={[mockNominee]} />
      </MockedProvider>
    );

    await waitFor(() => {
      expect(baseElement.querySelector(".Polaris-Sheet")).toMatchSnapshot();
    });
  });

  it("displays the correct number of nominations", async () => {
    const { baseElement } = render(
      <MockedProvider mocks={mocks}>
        <Sidebar sheetActive={true} nominees={[mockNominee]} />
      </MockedProvider>
    );

    await waitFor(() => {
      expect(
        baseElement.querySelectorAll(".Polaris-ResourceItem")
      ).toHaveLength(1);
    });
  });
});
