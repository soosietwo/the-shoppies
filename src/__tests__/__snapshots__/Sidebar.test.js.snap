// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar renders and matches snapshot 1`] = `
<div
  class="Polaris-Sheet"
  role="dialog"
  tabindex="-1"
>
  <div
    style="display: flex; flex-direction: column; height: 100%;"
  >
    <div
      style="align-items: center; border-bottom: 1px solid #DFE3E8; display: flex; justify-content: space-between; padding: 1.6rem; width: 100%;"
    >
      <h2
        class="Polaris-Heading"
      >
        Your nominations
      </h2>
      <button
        aria-label="Cancel"
        class="Polaris-Button Polaris-Button--plain Polaris-Button--iconOnly"
        type="button"
      >
        <span
          class="Polaris-Button__Content"
        >
          <span
            class="Polaris-Button__Icon"
          >
            <span
              class="Polaris-Icon"
            >
              <svg
                aria-hidden="true"
                class="Polaris-Icon__Svg"
                focusable="false"
                viewBox="0 0 20 20"
              >
                <path
                  d="M11.414 10l6.293-6.293a1 1 0 1 0-1.414-1.414L10 8.586 3.707 2.293a1 1 0 0 0-1.414 1.414L8.586 10l-6.293 6.293a1 1 0 1 0 1.414 1.414L10 11.414l6.293 6.293A.998.998 0 0 0 18 17a.999.999 0 0 0-.293-.707L11.414 10z"
                />
              </svg>
            </span>
          </span>
        </span>
      </button>
    </div>
    <div
      class="Polaris-Scrollable Polaris-Scrollable--vertical"
      data-polaris-scrollable="true"
      style="height: 100%;"
    >
      <div
        class="Polaris-ResourceList__ResourceListWrapper"
      >
        <ul
          aria-live="polite"
          class="Polaris-ResourceList"
        >
          <li
            class="Polaris-ResourceItem__ListItem"
          >
            <div
              class="Polaris-ResourceItem__ItemWrapper"
            >
              <div
                class="Polaris-ResourceItem"
              >
                <button
                  aria-label="View details for test"
                  class="Polaris-ResourceItem__Button"
                  tabindex="0"
                />
                <div
                  class="Polaris-ResourceItem__Container Polaris-ResourceItem--alignmentCenter"
                  id="132"
                >
                  <div
                    class="Polaris-ResourceItem__Owned"
                  >
                    <div
                      class="Polaris-ResourceItem__Media"
                    >
                      <img
                        alt="Poster for test"
                        src="123"
                        style="max-width: 100px;"
                      />
                    </div>
                  </div>
                  <div
                    class="Polaris-ResourceItem__Content"
                  >
                    <div
                      class="Polaris-TextContainer"
                    >
                      <h2
                        class="Polaris-Heading"
                      >
                        test
                      </h2>
                      <h3
                        class="Polaris-Subheading"
                      >
                        1234
                      </h3>
                      <button
                        class="Polaris-Button Polaris-Button--destructive"
                        type="button"
                      >
                        <span
                          class="Polaris-Button__Content"
                        >
                          <span
                            class="Polaris-Button__Text"
                          >
                            Remove
                          </span>
                        </span>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </li>
        </ul>
      </div>
    </div>
    <div
      style="align-items: center; border-top: 1px solid #DFE3E8; display: flex; justify-content: space-between; padding: 1.6rem; width: 100%;"
    >
      <button
        class="Polaris-Button"
        type="button"
      >
        <span
          class="Polaris-Button__Content"
        >
          <span
            class="Polaris-Button__Text"
          >
            Cancel
          </span>
        </span>
      </button>
      <button
        class="Polaris-Button Polaris-Button--primary Polaris-Button--disabled"
        disabled=""
        type="button"
      >
        <span
          class="Polaris-Button__Content"
        >
          <span
            class="Polaris-Button__Text"
          >
            Submit Nominees
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;
